File: android/app/src/main/kotlin/com/timeless/app/auth/TokenManager.kt
package com.timeless.app.auth

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.timeless.app.core.config.AppConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Manages authentication tokens with secure storage and automatic refresh
 */
class TokenManager(private val context: Context) {

    private val prefs by lazy {
        EncryptedSharedPreferences.create(
            context,
            PREFS_NAME,
            MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build(),
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }

    private val refreshMutex = Mutex()
    private var cachedTokens: AuthTokens? = null

    // MARK: - Public API

    /**
     * Get a valid access token, refreshing if necessary
     */
    suspend fun getAccessToken(): String? {
        val tokens = getTokens() ?: return null

        // Check if token is expired or about to expire
        if (isTokenExpired(tokens)) {
            return refreshTokens(tokens.refreshToken)?.accessToken
        }

        return tokens.accessToken
    }

    /**
     * Save tokens after successful authentication
     */
    fun saveTokens(tokens: AuthTokens) {
        prefs.edit()
            .putString(KEY_ACCESS_TOKEN, tokens.accessToken)
            .putString(KEY_ID_TOKEN, tokens.idToken)
            .putString(KEY_REFRESH_TOKEN, tokens.refreshToken)
            .putLong(KEY_EXPIRES_AT, tokens.expiresAt)
            .apply()
        cachedTokens = tokens
    }

    /**
     * Get current tokens from cache or storage
     */
    fun getTokens(): AuthTokens? {
        cachedTokens?.let { return it }

        val accessToken = prefs.getString(KEY_ACCESS_TOKEN, null) ?: return null

        val tokens = AuthTokens(
            accessToken = accessToken,
            idToken = prefs.getString(KEY_ID_TOKEN, "") ?: "",
            refreshToken = prefs.getString(KEY_REFRESH_TOKEN, "") ?: "",
            expiresAt = prefs.getLong(KEY_EXPIRES_AT, 0)
        )
        cachedTokens = tokens
        return tokens
    }

    /**
     * Clear all tokens (sign out)
     */
    fun clearTokens() {
        prefs.edit().clear().apply()
        cachedTokens = null
    }

    /**
     * Check if user has valid tokens
     */
    fun hasValidTokens(): Boolean {
        val tokens = getTokens() ?: return false
        return !isTokenExpired(tokens)
    }

    // MARK: - Token Refresh

    private fun isTokenExpired(tokens: AuthTokens): Boolean {
        val now = System.currentTimeMillis()
        return tokens.expiresAt - AppConfig.TOKEN_REFRESH_BUFFER_MS <= now
    }

    /**
     * Refresh tokens using the refresh token
     * Uses mutex to prevent multiple simultaneous refreshes
     */
    private suspend fun refreshTokens(refreshToken: String): AuthTokens? {
        return refreshMutex.withLock {
            // Double-check if another coroutine already refreshed
            cachedTokens?.let { cached ->
                if (!isTokenExpired(cached)) {
                    return@withLock cached
                }
            }

            performTokenRefresh(refreshToken)
        }
    }

    private suspend fun performTokenRefresh(refreshToken: String): AuthTokens? {
        return withContext(Dispatchers.IO) {
            try {
                val url = URL("https://${AppConfig.cognitoDomain}/oauth2/token")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
                connection.connectTimeout = (AppConfig.NETWORK_TIMEOUT_SECONDS * 1000).toInt()
                connection.readTimeout = (AppConfig.NETWORK_TIMEOUT_SECONDS * 1000).toInt()
                connection.doOutput = true

                val body = "grant_type=refresh_token" +
                        "&client_id=${AppConfig.cognitoClientId}" +
                        "&refresh_token=$refreshToken"

                connection.outputStream.use { it.write(body.toByteArray()) }

                if (connection.responseCode == 200) {
                    val responseText = connection.inputStream.bufferedReader().readText()
                    val json = JSONObject(responseText)

                    val tokens = AuthTokens(
                        accessToken = json.getString("access_token"),
                        idToken = json.getString("id_token"),
                        refreshToken = json.optString("refresh_token", refreshToken),
                        expiresAt = System.currentTimeMillis() + json.getLong("expires_in") * 1000
                    )

                    saveTokens(tokens)
                    tokens
                } else {
                    null
                }
            } catch (e: Exception) {
                null
            }
        }
    }

    companion object {
        private const val PREFS_NAME = "timeless_auth"
        private const val KEY_ACCESS_TOKEN = "access_token"
        private const val KEY_ID_TOKEN = "id_token"
        private const val KEY_REFRESH_TOKEN = "refresh_token"
        private const val KEY_EXPIRES_AT = "expires_at"

        @Volatile
        private var instance: TokenManager? = null

        fun getInstance(context: Context): TokenManager {
            return instance ?: synchronized(this) {
                instance ?: TokenManager(context.applicationContext).also { instance = it }
            }
        }
    }
}

/**
 * Authentication tokens data class
 */
data class AuthTokens(
    val accessToken: String,
    val idToken: String,
    val refreshToken: String,
    val expiresAt: Long
) {
    val isExpired: Boolean
        get() = System.currentTimeMillis() >= expiresAt - AppConfig.TOKEN_REFRESH_BUFFER_MS
}

File: ios/TimelessApp/TimelessApp/Features/Auth/TokenManager.swift
import Foundation

/// Manages authentication tokens with secure storage and automatic refresh
actor TokenManager {
    static let shared = TokenManager()

    private let keychain = KeychainService.shared
    private let config = CognitoConfig.shared

    /// Current tokens, if available
    private var cachedTokens: AuthTokens?

    /// Lock to prevent multiple simultaneous refresh operations
    private var isRefreshing = false

    /// Callbacks waiting for refresh to complete
    private var refreshCallbacks: [(Result<AuthTokens, Error>) -> Void] = []

    private init() {}

    // MARK: - Public API

    /// Get a valid access token, refreshing if necessary
    func getAccessToken() async throws -> String {
        // Try to get cached tokens
        let tokens = try await getTokens()

        guard let tokens = tokens else {
            throw TokenError.notAuthenticated
        }

        // Check if token is expired or about to expire
        if tokens.isAccessTokenExpired {
            let refreshedTokens = try await refreshTokens(tokens.refreshToken)
            return refreshedTokens.accessToken
        }

        return tokens.accessToken
    }

    /// Save tokens after successful authentication
    func saveTokens(_ tokens: AuthTokens) async throws {
        try await keychain.saveTokens(tokens)
        cachedTokens = tokens
    }

    /// Get current tokens from cache or Keychain
    func getTokens() async throws -> AuthTokens? {
        if let cached = cachedTokens {
            return cached
        }

        let tokens = try await keychain.getTokens()
        cachedTokens = tokens
        return tokens
    }

    /// Clear all tokens (sign out)
    func clearTokens() async throws {
        try await keychain.deleteTokens()
        cachedTokens = nil
    }

    /// Check if user has valid tokens
    func hasValidTokens() async -> Bool {
        do {
            let tokens = try await getTokens()
            return tokens != nil && !(tokens?.isAccessTokenExpired ?? true)
        } catch {
            return false
        }
    }

    // MARK: - Token Refresh

    /// Refresh tokens using the refresh token
    private func refreshTokens(_ refreshToken: String) async throws -> AuthTokens {
        // If already refreshing, wait for the result
        if isRefreshing {
            return try await withCheckedThrowingContinuation { continuation in
                refreshCallbacks.append { result in
                    continuation.resume(with: result)
                }
            }
        }

        isRefreshing = true
        defer {
            isRefreshing = false
            refreshCallbacks.removeAll()
        }

        do {
            let tokens = try await performTokenRefresh(refreshToken)
            try await saveTokens(tokens)

            // Notify waiting callbacks
            for callback in refreshCallbacks {
                callback(.success(tokens))
            }

            return tokens
        } catch {
            // Notify waiting callbacks of failure
            for callback in refreshCallbacks {
                callback(.failure(error))
            }
            throw error
        }
    }

    /// Perform the actual token refresh HTTP request
    private func performTokenRefresh(_ refreshToken: String) async throws -> AuthTokens {
        let url = URL(string: "https://\(config.domain)/oauth2/token")!

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")

        let body = [
            "grant_type": "refresh_token",
            "client_id": config.clientId,
            "refresh_token": refreshToken
        ]
        request.httpBody = body
            .map { "\($0.key)=\($0.value)" }
            .joined(separator: "&")
            .data(using: .utf8)

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw TokenError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            throw TokenError.refreshFailed(httpResponse.statusCode)
        }

        let tokenResponse = try JSONDecoder().decode(TokenResponse.self, from: data)

        return AuthTokens(
            accessToken: tokenResponse.access_token,
            idToken: tokenResponse.id_token,
            refreshToken: tokenResponse.refresh_token ?? refreshToken,
            expiresAt: Date().addingTimeInterval(TimeInterval(tokenResponse.expires_in))
        )
    }
}

// MARK: - Supporting Types

enum TokenError: LocalizedError {
    case notAuthenticated
    case invalidResponse
    case refreshFailed(Int)
    case expired

    var errorDescription: String? {
        switch self {
        case .notAuthenticated:
            return "Not authenticated"
        case .invalidResponse:
            return "Invalid response from server"
        case .refreshFailed(let code):
            return "Token refresh failed with status \(code)"
        case .expired:
            return "Session expired, please sign in again"
        }
    }
}

private struct TokenResponse: Codable {
    let access_token: String
    let id_token: String
    let refresh_token: String?
    let expires_in: Int
    let token_type: String
}
